core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1761,6 +1776,59 @@ private[spark] class BlockManager(
     blocksToRemove.size
   }
 
+  def decommissionBlockManager(): Unit = {
+    if (!blockManagerDecommissioning) {
+      logInfo("Starting block manager decommissioning process")
+      blockManagerDecommissioning = true
+      decommissionManager = Some(new BlockManagerDecommissionManager)
```

Comment:
I guess I'm not sure why were doing this in this manner? It seems like if we did this as a blocking call and returned true when we were done we could move on to a next phase of decommissioning once the blocks were replicated in the driver?

Reply:
@holdenk I have added the overall design for this change. Please review the same and provide your suggestions/feedback.

Reply:
Resolving this as we have another comment thread going on about why a thread is needed instead of blocking call.